packages/std/src/math/fraction.rs

```diff
@@ -12,3 +12,57 @@ pub trait Fraction<T>: Sized {
     /// If `p` is zero, None is returned.
     fn inv(&self) -> Option<Self>;
 }
+
+impl<T: Clone> Fraction<T> for (T, T) {
+    fn numerator(&self) -> T {
+        self.0.clone()
+    }
+
+    fn denominator(&self) -> T {
+        self.1.clone()
+    }
+
+    fn inv(&self) -> Option<Self> {
+        Some((self.1.clone(), self.0.clone()))
+    }
+}
+
+#[macro_export]
+macro_rules! impl_mul_fraction {
+    ($UintBase:ident, $UintLarger:ident) => {
+        impl $UintBase {
+            pub fn checked_mul_floored<F: Fraction<T>, T: Into<$UintBase>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionError> {
+                let res = self
+                    .full_mul(rhs.numerator().into())
+                    .checked_div($UintLarger::from(rhs.denominator().into()))?;
```

Comment:
This part should work with .into().into()

```suggestion
               let divisor = rhs.denominator().into(); // guaranteed to be $UintBase now
                let res = self
                    .full_mul(rhs.numerator().into())
                    .checked_div(divisor.into())?;
```